python/ray/data/_internal/execution/streaming_executor_state.py

```diff
@@ -359,17 +361,30 @@ def process_completed_tasks(
             fetch_local=False,
             timeout=0.1,
         )
+
+        # Organize tasks by the operator they belong to, and sort them by task index.
+        # So that we'll process them in a deterministic order.
+        # This is because some backpressure policies (e.g.,
+        # StreamingOutputBackpressurePolicy) may limit the number of blocks to read
+        # per operator. In this case, we want to have fewer tasks finish quickly and
+        # yield resources, instead of having all tasks output blocks together.
```

Comment:
Nice.

Can we write a test for this? Should be possible with the new task metrics testing utils.

Reply:
The new test case already covers this. Without this change, all tasks will be making progress together, and the disk spilling amount will be much higher.